swiftide-integrations/src/openai/mod.rs

```diff
@@ -211,6 +213,66 @@ impl<C: async_openai::config::Config + Default> GenericOpenAI<C> {
     }
 }
 
+pub fn openai_error_to_language_model_error(e: OpenAIError) -> LanguageModelError {
+    match e {
+        OpenAIError::ApiError(api_error) => {
+            // If the response is an ApiError, it could be a context length exceeded error
+            if api_error.code == Some("context_length_exceeded".to_string()) {
+                LanguageModelError::ContextLengthExceeded(OpenAIError::ApiError(api_error).into())
+            } else {
+                tracing::error!("OpenAI API Error: {:?}", api_error);
+                LanguageModelError::PermanentError(OpenAIError::ApiError(api_error).into())
+            }
+        }
+        OpenAIError::Reqwest(e) => {
+            if let Some(status) = e.status() {
+                // If the response code is 429 it could either be a TransientError or a
+                // PermanentError depending on the message, if it contains the word
+                // quota, it should be a PermanentError otherwise it should
+                // be a TransientError.
+                // If the response code is any other 4xx it should be a PermanentError.
+                if status.as_u16() == 429 && !e.to_string().contains("quota") {
+                    LanguageModelError::TransientError(e.into())
+                } else if status.is_client_error() {
+                    tracing::error!("OpenAI API Client Error: {:?}", e);
+                    LanguageModelError::PermanentError(e.into())
+                } else if status.is_server_error() {
+                    tracing::warn!("OpenAI API Server Error: {:?}", e);
+                    LanguageModelError::TransientError(e.into())
+                } else {
+                    tracing::error!("Unexpected OpenAI Error: {:?}, error: {:?}", status, e);
+                    LanguageModelError::PermanentError(e.into())
+                }
+            } else {
+                // making the request failed for some other reason, probably recoverable
+                tracing::error!("Unexpected OpenAI Reqwest Error: {:?}", e);
+                LanguageModelError::TransientError(e.into())
+            }
+        }
+        OpenAIError::JSONDeserialize(e) => {
+            // OpenAI generated a non-json response, probably a temporary problem on their side
+            tracing::error!("OpenAI response could not be deserialized: {:?}", e);
+            LanguageModelError::TransientError(e.into())
```

Comment:
This would be a cloudfare response then?

Reply:
Yeah most likely